File: acg-android/app/src/main/java/com/aicraftspeopleguild/acg/ui/SettingsScreen.kt
package com.aicraftspeopleguild.acg.ui

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.aicraftspeopleguild.acg.data.Mqtt
import com.aicraftspeopleguild.acg.data.SyncEngine
import com.aicraftspeopleguild.acg.data.TagDb
import com.aicraftspeopleguild.acg.data.TokenStore
import com.aicraftspeopleguild.acg.udt.UdtRegistry
import com.aicraftspeopleguild.acg.ui.organism.CatalogSection
import kotlinx.coroutines.launch

@Composable
fun SettingsScreen(
    tokens: TokenStore, tagDb: TagDb, sync: SyncEngine,
    registry: UdtRegistry,
    modifier: Modifier = Modifier,
) {
    var input by remember { mutableStateOf(tokens.getToken().orEmpty()) }
    var status by remember { mutableStateOf<String?>(null) }

    Column(
        modifier.verticalScroll(rememberScrollState()).padding(20.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Text("GitHub token", style = MaterialTheme.typography.titleMedium)
        Text(
            "Required for tag_write + cmd_action. Stored in EncryptedSharedPreferences, excluded from cloud backup.",
            fontSize = 11.sp,
            color = MaterialTheme.colorScheme.onBackground.copy(alpha = 0.6f),
        )
        OutlinedTextField(
            value = input,
            onValueChange = { input = it },
            label = { Text("GitHub PAT (repo scope)") },
            singleLine = true,
            visualTransformation = PasswordVisualTransformation(),
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Password),
            modifier = Modifier.fillMaxWidth(),
        )
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            Button(onClick = {
                tokens.setToken(input.trim().takeIf { it.isNotEmpty() })
                status = if (tokens.hasToken()) "✓ saved" else "cleared"
            }) { Text("Save") }
            OutlinedButton(onClick = {
                input = ""; tokens.setToken(null); status = "cleared"
            }) { Text("Clear") }
        }
        status?.let { Text(it, color = MaterialTheme.colorScheme.primary) }

        Spacer(Modifier.height(12.dp))
        HorizontalDivider()
        Text("Local tag.db", style = MaterialTheme.typography.titleMedium)
        val scope = rememberCoroutineScope()
        var syncStatus by remember { mutableStateOf<String?>(null) }
        var busy by remember { mutableStateOf(false) }
        Text("tags cached: ${tagDb.countTags()}  ·  mqtt subs: ${Mqtt.subCount()}  ·  retained: ${Mqtt.retainedCount()}",
             fontSize = 11.sp,
             color = MaterialTheme.colorScheme.onBackground.copy(alpha = 0.6f))
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            Button(onClick = {
                scope.launch {
                    busy = true; syncStatus = null
                    try {
                        val r = sync.pullAllTags()
                        syncStatus = "✓ ${r.tagsPulled} tags · ${r.errors} errors · ${r.durationMs}ms"
                    } catch (e: Exception) { syncStatus = "✗ ${e.message}" }
                    busy = false
                }
            }, enabled = !busy) { Text("Sync now") }
        }
        syncStatus?.let { Text(it, fontSize = 11.sp,
            color = if (it.startsWith("✓")) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.error) }

        Spacer(Modifier.height(12.dp))
        HorizontalDivider()
        CatalogSection(registry)

        Spacer(Modifier.height(12.dp))
        HorizontalDivider()
        Text("About", style = MaterialTheme.typography.titleMedium)
        Text("ACG Android 0.3.0 · L0 sensor plant · local tag.db · Mqtt bus · UDT catalog.", fontSize = 12.sp)
        Text("Fork: teslasolar/aicraftspeopleguild.github.io  ·  watches origin too", fontSize = 12.sp,
             color = MaterialTheme.colorScheme.onBackground.copy(alpha = 0.6f))
    }
}

File: phone/acg-ios/ACG/Screens/SettingsScreen.swift
import SwiftUI

struct SettingsScreen: View {
    @ObservedObject var world: World
    @State private var input: String = ""
    @State private var saveStatus: String?
    @State private var expanded = Set<String>()

    var body: some View {
        Form {
            Section("GitHub token") {
                Text("Required for tag_write + cmd_action. Keychain-stored.").font(.caption).foregroundStyle(.secondary)
                SecureField("GitHub PAT (repo scope)", text: $input)
                HStack {
                    Button("Save") {
                        world.tokens.save(input.trimmingCharacters(in: .whitespacesAndNewlines))
                        saveStatus = world.tokens.hasToken ? "✓ saved" : "cleared"
                    }
                    Button("Clear", role: .destructive) {
                        input = ""; world.tokens.delete(); saveStatus = "cleared"
                    }
                    if let s = saveStatus { Text(s).foregroundStyle(.green) }
                }
            }

            Section("Local tag.db") {
                Text("tags cached: \(world.tagDb.count())  ·  mqtt subs: \(world.mqtt.subCount)  ·  retained: \(world.mqtt.retainedCount)")
                    .font(.caption).foregroundStyle(.secondary)
                Button("Sync now") { Task { _ = await world.sync.pullAllTags() } }
            }

            Section("UDT CATALOG · \(world.registry.instances.count) instances / \(world.registry.typesPresent.count) types") {
                ForEach(Array(world.registry.countByType.keys).sorted(), id: \.self) { ty in
                    DisclosureGroup(isExpanded: Binding(
                        get: { expanded.contains(ty) },
                        set: { on in if on { expanded.insert(ty) } else { expanded.remove(ty) } }
                    )) {
                        ForEach(world.registry.instances(of: ty), id: \.instance) { inst in
                            VStack(alignment: .leading, spacing: 2) {
                                Text(inst.instance).font(.system(.callout, design: .monospaced).bold())
                                if let id = inst.str("id") {
                                    Text("id \(id)").font(.caption2).foregroundStyle(.secondary)
                                }
                                let pubs = inst.strList("publishes"), subs = inst.strList("subscribes")
                                if !pubs.isEmpty || !subs.isEmpty {
                                    Text((pubs.isEmpty ? "" : "pub " + pubs.joined(separator: ",")) +
                                         (subs.isEmpty ? "" : "  ·  sub " + subs.joined(separator: ",")))
                                        .font(.caption2).foregroundStyle(.secondary)
                                }
                            }.padding(.vertical, 2)
                        }
                    } label: {
                        HStack {
                            Text(ty).font(.system(.body, design: .monospaced).bold())
                            Spacer()
                            Text("\(world.registry.countByType[ty] ?? 0)")
                                .font(.system(.caption, design: .monospaced)).foregroundStyle(.secondary)
                        }
                    }
                }
            }

            Section("About") {
                Text("ACG iOS 0.1.0 · SwiftUI · Keychain PAT · Mqtt bus · UDT catalog.").font(.footnote)
                Text("Fork: teslasolar/aicraftspeopleguild.github.io  ·  watches origin too")
                    .font(.footnote).foregroundStyle(.secondary)
            }
        }
    }
}
